### DISCUSSION TITLE
SSR caching between requests? No, but global state will!

### DISCUSSION THREAD

**Discussion by fairbairn** [vercel/next.js#15954] [Help] [answered]
I'm a bit confused how this is happening... we're using next-redux-wrapper, but that's really irrelevant for the question..

We have a page, that is server side rendered...
If we load it up via localhost:3000 in the browser and hit refresh (refreshing the browser), we see the page.json file being requested over the Network...
Somehow in this page.json response, it is getting redux hydrating from some cached version on the server...
To test our theory, we created a counter on the SSR page, in the getServerSideProps function, we dispatch to increment a counter (in redux), then grab that value out of redux as a prop to pass to the page, which is rendered...
No problem that works.

However, we can't get it NOT TO WORK.

If we go to the terminal, and curl the exact same URL, passing zero else (no headers, no cookies, nothing), the pageProps contains data that it should not have - it somehow has a cached version of the prior page props and returns a page with those values present...(in "__NEXT_DATA__")

Since getServerSideProps should know nothing, how on a command line curl is it returning a prop with a count that's sequential from the web page doing the incrementing??

I would think a curl would respond with the page at count = 1 (count = 0 upon new page, we increment by 1 server side, we render with 1).  Hitting curl should always return count = 1.

We can hit it from a curl and get 168, 169, 170, hit it from a browser, 171, 172, 173, hit it from curl, 174, 175.

If a google spider hit the page url, we would not expect it to appear as if it was remembering the state embedded on the page from one request to the next, but that seems to be what's happening...

The only place this information could reside is within the nextjs server system itself.  It must be remembering the NEXT_DATA page state between requests, but on an SSR page, our assumption is it should never do that, it should always render it upon request... it is responding per request with the page.json response, but that data, somehow, is not new.

We're testing this in production mode, so running after a build with npm start...

Is there some optimizing going on in 9.5 that we need to turn off?  Seems to me it maintains a cached version of those pageProps between server requests (regardless of the client making the request)...

Reason this is relevant is we actually have a cookie on the browser, that on a SSR, we use to perform an auto-login and then store secret credentials in props that we send back to the client to maintain...

Unfortunately, if we hit a running server from another machine or a terminal or an anonymous web session, it's returning those props from another former request....it's caching that data on an SSR response..

Any help or insight would be greatly appreciated.

**Comment by fairbairn** [comment]
So this was a tricky one - two things caused this issue...

The first was that the `store` that we were using with next-redux-wrapper was defined globally const, instead of internally within makeStore - that meant that nextjs was somehow remembering the cached global state chunk, even after redux mods, from request to request, and essentially the redux code was not re-initializing the store.  Now how that is even possible is beyond me, as I would think on an SSR (especially from curl) that there would be no `state` in which to render, but that's actually not the case...

So there's still a question as to how nextjs is caching that data (intelligently) across SSR requests, but the primary state scenario was caused because next-redux-wrapper was never re-initializing, it too picked up the cached state.

I'd love to know more how the actual SSR logic works, what is cached, how NEXT_DATA is handled in that regard, and if it's truly a clean request.

**Comment by fairbairn** [comment, marked as answer]
So in final follow up, the reason it was able to maintain a central state across requests from different clients, is because we had declared the redux store as global...hence when redux actions modified the state, that state was accessible by other unrelated requests (which actually opened up a security hole).

I guess it makes sense - in the context of a single process running node server, that global information is in one place, and it's not removed per request...

We use stateless web servers, like Elixir, Nginx, etc, and so it was confusing how state could be maintained across requests.

Now it makes sense.

It became clear once we started to stop and restart the server, at which point the state reset.

(done and done)

**Reply by doubleniki** [reply]
So, what was the solution? Server stop -> start? :)

**Reply by kumarldh** [reply]
Starting and stoppping the server every now and then doesn't sound a correct solution...does this mean redux store was acting like "application cache"?

**Reply by doubleniki** [reply]
@fairbairn could you please share your solution with us? :)

**Comment by williamespindola** [comment]
I thint that I have the same issue here. Some how in de SSR something keep the store as the first fill but in the client the state is fresh with each changes. Only after a `pm2 [service] restart` a new state and only at the first request this state is saved and not change intil a new restart from the pm2.
I can see this problem cleary because I have some cookie params that I send with requests, some console logs at the call of request I can see the new data, but on the api only the old param arraives.

**Comment by fairbairn** [comment]
So as stated, when we originally designed our solution, we had our `store` as a global const in the module, which meant that it was permanent.  It would essentially persist and hold that state across pages because the node server had it in memory.

We altered that configuration, so the store was instead created by makeStore..and we built a wrapper that we use in `_app.js` to wrap the entire application...

Since the store is now loaded as a part of the App, not globally in memory server side, it would essentially start from scratch on a full reload of the page (full refresh, not a click which is a next.js page flipping operation.

We built a `withApp` HOC that wraps our core logic... it looks a bit like this...

```
const makeStore = () => {
    let state = {};
    return configureStore(Immutable(state));
};

export const appWrapper = createWrapper(makeStore, {debug: false});

const select = (state, props) => {
    return {
        state,
        dispatch: props.dispatch,
    };
};

export const withApp = (app) => appWrapper.withRedux(connect(select)(app));

```

and then in our _app.js, we always configure like this..

```
export default withApp(App);
```

your App is simply something like..

```
import {withApp} from "wrapper";

function App(props) {
    const {Component, pageProps, state, dispatch} = props;

    const renderLayout = (props) => {
        return (
                <Component {...props} />
        );
    };

    return <Adapter {...props} {...pageProps} render={renderLayout} />;
}

export default withApp(App);
```

Please note, our example is a bit more complicated, we have yet another HOC called Adapter that does some things like injects the router, the route name, the route path, host and some top level functions we want throughout.

But generally, this handles the redux scenario.  The trick is makeStore.  With this foundation, we can easily add logged in states, serialization, and other features all within wrapper.  

Hope this helps.

**Reply by doubleniki** [reply]
Many thanks for the solution example! Have a nice day. 👏

**Reply by kumarldh** [reply]
@fairbairn thanks for the detailed solution. I may sound dummy for asking this. Please bear with me and help me understand following, this is confusing me

>Since the store is now loaded as a part of the App, not globally in memory server side, it would essentially start from scratch on a full reload of the page (full refresh, not a click which is a next.js page flipping operation.

I have just one excuse, English isn't my first or even 2nd language. 😁

**Reply by doubleniki** [reply]
@fairbairn I have a couple of additional questions. Do you use the redux store on the server side? Do you push anything to the store during SSR and then hydrate the client-side store with data from the server-side? Thanks.
